pkg/sql/sqlbase/table.go

```diff
@@ -1579,27 +1615,59 @@ func MarshalColumnValue(col ColumnDescriptor, val parser.Datum) (roachpb.Value,
 		val.ResolvedType(), col.Type.SemanticType, col.Name)
 }
 
+func encodeArrayHeader(h arrayHeader, buf []byte) ([]byte, error) {
+	// The byte we append here is formatted as follows:
+	// * The high 4 bits encode the number of dimensions in the array.
+	// * The lowest bit records whether the array contains NULLs.
+	// * The remaining bits are reserved.
+	headerByte := h.numDimensions << 4
+	if h.hasNulls {
+		headerByte = headerByte | 1
+	}
+	buf = append(buf, byte(headerByte))
+	buf = encoding.EncodeValueTag(buf, encoding.NoColumnID, h.elementType)
+	buf = encoding.EncodeNonsortingUvarint(buf, h.length)
+	return buf, nil
+}
+
 func encodeArray(d *parser.DArray, scratch []byte) ([]byte, error) {
 	if err := d.Validate(); err != nil {
 		return scratch, err
 	}
 	scratch = scratch[0:0]
-	// The byte we append here is formatted as follows:
-	// * The high 4 bits encode the number of dimensions in the array.
-	// * The lowest bit records whether the array contains NULLs.
-	// * The remaining bits are reserved.
-	scratch = append(scratch, 1<<4)
-	encType, err := parserTypeToEncodingType(d.ParamTyp)
+	elementType, err := parserTypeToEncodingType(d.ParamTyp)
+	if err != nil {
+		return nil, err
+	}
+	header := arrayHeader{
+		hasNulls: d.HasNulls,
+		// TODO(justin): support multiple dimensions.
+		numDimensions: 1,
+		elementType:   elementType,
+		length:        uint64(d.Len()),
+		// We don't encode the NULL bitmap this way because we do it in lockstep with the
```

Comment:
Clarify what "this way" means here. Also Go initializes fields to nil by default.